Adding non-linear integrator take-back-half (TBH) feature to PID

CircuitGuy opened the pull request:
### How to Use
* Off by default - In the CLI _set iterm_tbh = RPY_ {options are OFF, RPY, or RP}
* I term can likely be increased ~1.5x with no negative side effects

### What it does
* The I term of the quadcopter will make some small non-linear "jumps" when error crosses zero.  
* At steady-state, they are negligible.  
* However, during high error periods, the jumps can become much longer.
* The jumps "look" a little like D-term, but will not ring in the same way
* The jumps do cause sudden motor acceleration, just like D-term.

### Figure and logs
* TBH = RPY on left
* TBH = OFF on right.
* PID are defaults to the BF dev defaults; including the new FF term for fair comparisons.
![image](https://user-images.githubusercontent.com/1209150/43812342-79bbea48-9a86-11e8-8a70-76efaa51c3e2.png)
[Quad Logs 5 Inch Cham Ti TBH 2018.08.07.zip](https://github.com/betaflight/betaflight/files/2268690/Quad.Logs.5.Inch.Cham.Ti.TBH.2018.08.07.zip)


### How it's different from other i-term features
I term relax is the normal feature used to modify the I-term of the quadcopter.  However, I-term relax effectively disables the integrator during high-speed maneuvers.  This instead allows the I-term to over-accumulate but then jumps back halfway towards the last known-good position.

The TBH does not disable the saturation features of BetaFlight, so that non-linearity is kept intact.

### Theory of operation
This is the take-back-half algorithm (TBH).  This takes a different approach and adds a non-linear active feature to the integrator; similar to a saturation effect or I-term relax.  Imagine the optimal controller to accelerate a vehicle from a dead stop to highway speed as fast as possible.  As a human former-teenager, we know the best way is to put the gas pedal to 100% until we're close to our desired speed, then suddenly jump back.  This is a non-linear response that a traditional PID controller can't really do.  If it reacts quickly enough to jump towards 100% input, it will tend to not back off enough either once you reach speed.  However, this algorithm does exactly that - once you reach your desired setpoint, it removes half of the accumulated integrator error.  When you reach it again (presumably you're not perfect), it makes consecutively smaller jumps.

If we turn our I-term up too high, normally because the integrator has been 'pushing' the quadcopter back to zero error, once at zero error, the integrator has some additional wind-up.  In order to eliminate this, we normally tune the I term to be ~ critically damped.  However, if we reset the I term to a better location (which we'll define as halfway between present value and the value when the error was last ~0), we can make it more responsive without the overshoot penalty.  Additional information is used to reduce the phase delay over a linear controller.  The value of half is important because it can be shown to converge to unity when summed to infinity against an otherwise unstable plant (1/2 + 1/4 + 1/8... = 1); in our case stability (but not good performance!) is possible with any Ki.

### Known issues/limitations
* This is my first PR - coding is a hobby, not a job.  So... constructive feedback welcome
* Did not add to configurator app
* I don't get enough flying time! Having trouble gathering good data for this PR; I'd love it if anybody is willing to offer performance feedback.  Tested with a friend on his 3", his 5", and my 5" (logs provided for mine only)

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -990,6 +991,22 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+                // the I term to non-linearly jump to half-way between the current
+                // accumulated error and the accumulated error at the last zero
+                // crossing.
+                const float ThisSign = copysign(1.0, itermErrorRate); //-1 or 1
```
`1.0` type is double, should be `1.0f`.
Also, we use `camelCase` for variables

CircuitGuy replied:
Good catch :)

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -990,6 +991,22 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+                if (ThisSign != pidData[axis].errorSignPrevTBH) { 
+                    const float lastCrossIValue = pidData[axis].IPrevTBH;
+                    pidData[axis].IPrevTBH = pidData[axis].I;
+                    pidData[axis].I = (pidData[axis].I + lastCrossIValue)*0.5; //TBH
```
Also this

CircuitGuy commented:
@McGiverGim 
Attached.  Log is fairly short, just bobbling around in prop wash with some sharp rolls in pitch and yaw.  Tune is a little oscillatory here (really needs a little d-axis or reduced P), especially roll.  For a clear demonstration, I'd suggest turning on PID error and PID I for all three axes.  Around 34.5 seconds is a good (albeit poorly tuned)

[LOG00139.zip](https://github.com/betaflight/betaflight/files/2239211/LOG00139.zip)

 demonstration of what the algorithm is doing.

ctzsnooze commented:
![pr6487](https://user-images.githubusercontent.com/11737748/43371069-b918ae0c-93cd-11e8-9c39-a24c097f303d.jpg)

Definitely not tuned.  Need to see a tuned log without the slow unstable wobble, and with some 360 degree flips, to see what it's capable of.  For some reason, no D in the blackbox log explorer.  Behaving like a laggy PI (mostly P) controller as far as I can see.

Hopefully this log is not indicative of a good concept.  Let's see how it behaves once optimised.

McGiverGim commented:
I think this is the objective. If I have understood the explanation he has removed D and uses a modified I.

CircuitGuy commented:
Yeah, this particular quad has a fairly springy frame. Given any tune with a linear PID, turning on the TBH *should* make it better. It will add about 15* phase margin without slowing down the step response.

RipperDrone commented:
1+1/2+1/4...1/2*n= 1 for n --> infinite? Got the sequence wrong, I guess. Should start at n=2, I suppose...

Hope the coding is still correct... 🙂

McGiverGim commented:
I've been looking at the Blackbox file and what I see is something similar to @ctzsnooze:
- The P is constantly overshooting, producing a wobble, as expected without using any of D. The wobble is small, but this log is only a hover, I expect a lot more in agressive flight.
- The I is more linear, and doing some little jumps trying to help P to stop, but it seems not enough in any case.

And this only in a hover log. Without almost any movement.

I think we need a sample of a real tuned quad, and a real flight with fast movements to see if this solution can be real used. At this moment it seems that not.

McGiverGim commented:
@CircuitGuy one question, why is the TBH related to the zero D?

I was thinking to add a feature to accelerate the I Term only when decreasing. My idea was make it lineal but faster. Your code is doing it in a not lineal way but the idea is the same (reduce the I Term when decreasing). But I don't understand the relation between D and this TBH I Term. 

The P and the I both push in the same direction when there is an error, but I don't see any force trying to stop the movement BEFORE arriving to zero, so it will be overshoot always.

As I say, maybe is a good idea this non-linear I Term integrator, but I can't see it as a substitute of the D-Term.

ledvinap commented:
@CircuitGuy : Can you test in in level/horizon mode too?

IMO introducing nonlinearity (and departing from proven PID algorithm) is only making maintenance harder. It makes sense if it provides visibly better results, but not in other cases.

Multiwii was full of heuristics, usually of unknown origin and intent. Baseflight and cleanflight removed a lot of that, but it is now creeping back ...

CircuitGuy commented on src/main/flight/pid.c:
```diff
@@ -990,6 +991,22 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+                // crossing.
+                const float thisSign = copysign(1.0f, itermErrorRate); //-1 or 1
+                if (thisSign != pidData[axis].errorSignPrevTBH) { 
+                    const float lastCrossIValue = pidData[axis].IPrevTBH;
```
TBH is meant to be an acronym.  Given that, is "Tbh" still preferable?  Makes it less readable IMHO, but I have no problem updating it to conform to the styleguide

mikeller replied:
Generally, acronyms are used in camelCase by only capitalising the first letter - this allows for them to be used at the start of lower cased names without looking ugly, i.e. `tbhLimit` instead of `tBHLimit`.

frozenskys commented:
Have been looking at this algorithm and it looks like it was designed for pure-integration (I term only) controllers mainly aimed at bang-bang (on-off) sensors https://www.edn.com/design/sensors/4460914/Generate-proportional-feedback-from-a-bang-bang-sensor-with-a-take-back-half-controller- so I'm not sure that it will work ideally with gyros...

CircuitGuy commented:
Updated the PR with all comments.

* Not requested, but I made it more consistent with the iterm_relax in code and settings.  This should also eliminate some of @ledvinap concern's
* @RipperDrone  Good catch 👍 
* Removed the D-term discussion and made the pull request more readable.  D-term is an end goal of mine unrelated to this

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -224,6 +225,11 @@ static FAST_RAM_ZERO_INIT uint8_t itermRelaxType;
 
+#if defined(USE_TBH)
+static FAST_RAM_ZERO_INIT float iPrevTBH[XYZ_AXIS_COUNT];
+static FAST_RAM_ZERO_INIT float errorSignPrevTBH[XYZ_AXIS_COUNT];
```
`int8_t` or 'bool',  would be enough and save some ram
(`bool thisSign  = itermErrorRate > 0`)

ctzsnooze commented:
@CircuitGuy - any chance you could upload a log of normal flying, including fast forward flight and clean flips, with normal PID's, one with it off, one on?

CircuitGuy commented:
@ctzsnooze Agreed. I think. I'm trying to say that "disabling the integrator" is equivalent to halting the integration process. I.e. latching it. I might be missing something in iterm relax though, are you saying it lowers Ki without going all the way to zero? 

Either way the difference here is that TBH allows Ki to remain active during transients but then quickly reset itself. This algorithm works very well for plants that have very different large scale and small scale traits. I was trying to gloss over it, but the infinite series example I gave can be used when calculating the system's large scale stability Lyapunov function. Given that it converges, with most plants (quads included), this algorithm guarantees limited energy with any Ki, which is impossible with any linear controller combined with a plant that has delay. Put another way, it allows the integration to work to stabilize the plant, but bounds the energy. The iterm relax sets the gain low enough (or removes it? Still not sure) that it's stable even with the delay.

For logs, are you looking for something other than what's linked in the updated pull request? If you're looking for cleaner flips, I might have to find a better pilot. :) I totally agree that the log I posted in the comments isn't helpful for showing what the algorithm can do, but the updated main pull request ones were meant to answer that question.

McGiverGim commented:
Iterm Relax locks the I Term (it does not reset it) in fast moves (while others elements like P are doing the job).
The doubt with your code are the jumps. I tested with some pilots a jump to zero some time ago and it did not work right. Yours is a jump to a half, so maybe it's better, let's see.

CircuitGuy commented:
@McGiverGim Yeah, that makes sense.  I think the extra piece I'm not explaining is that it doesn't change Ki at all, and it doesn't jump halfway to zero.  

When the sign of the error changes, it takes away half of the integrated error (I term output).  This allows I to ramp up very quickly, but then shed all of the accumulated error once the setpoint is reached.  Another important piece; it doesn't go halfway to zero, it goes halfway to whatever the last zero-crossing accumulated error is.  I.e. you're flying around and I-term is steady-state 50, you make a rapid move, and your quad lags behind.  Say the integrator ends up accumulating all the way to 100 for this fast move.  Once you've hit the commanded rate again, it jumps to 75 (halfway back to 50).  Assume it was a bit too much, and the I begins falling and zero error his hit again at 40, TBH then jumps accumulated error to 45.  The first jump from 100-> 50 prevents a windup error, but I-term still added a lot to help you track the changes (storing energy in the quad's inertia), potentially much more than P, especially as you started approaching zero error.  Now that you're at steady-state, the TBH jumps should be negligible compared to the noise in your quad (i.e. I term does nothing at steady state except ~ hold its output).

DieHertz commented on src/main/flight/pid.c:
```diff
@@ -990,6 +996,22 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
+                // accumulated error and the accumulated error at the last zero
+                // crossing.
+                const bool thisSign = itermErrorRate > 0; //T==Pos, F==Neg or zero
+                if (thisSign != tbhErrorSignPrev[axis]) { 
```
I'd prefer storing `itermErrorRatePrev` rather than boolean for sign and then checking for  `itermErrorRate * itermErrorRatePrev < 0`

robert-b commented:
the idea is not bad and it works.
![image](https://user-images.githubusercontent.com/1392279/44810421-f57e2680-abd1-11e8-9978-a705abd50cc9.png)
i am using the same - but with a hard reset of the integrator value.
no bad behavior on a flip.
